Saving data

jonholmberg opened the pull request:
The app now saves all messages written to a temporary file on the internal storage, and added a button that creates an activity that shows all the previously written messages with a timestamp (i.e. it reads the contents of the above mentioned file). The file will remain on the internal storage when the app is closed, so that you can see messages you wrote last time you had the app open, and it wont be deleted until the app is uninstalled.

Added new icon

jonholmberg committed:
Added a feature that shows the message shown last time app was closed

jonholmberg committed:
The Show All Messages button now creates a new activity with a message

Nightly commit

Changed layout a bit in the fragent that displays the message

The Show all messages button now creates a new activity with a message

Added Host image

The app should know write all message to a temporary file

The Show all messages button now shows a message

jonholmberg committed:
Nightly commit

jonholmberg committed:
Button reference fix

Button reference fix

Random generated cange

jonholmberg committed:
Added previous message feature

Button reference fix

Button reference fix

Random generated cange

Added a feature that shows the message shown last time app was closed

jonholmberg committed:
Changed layout a bit in the fragent that displays the message

jonholmberg committed:
The Show all messages button now creates a new activity with a message

Added Host image

The app should know write all message to a temporary file

The Show all messages button now shows a message

jonholmberg committed:
Merge branch 'saving-data' of github.com:e-conomic/android-course into saving-data

jonholmberg committed:
Added a few logs in MainActivity

jonholmberg committed:
The ShowAllMessagesActivity now shows a message of any size

jonholmberg committed:
Fixed issue where the wrong message is read to the DisplayAllMessagesActivity

jonholmberg committed:
Moved the code that retrieves the written messages to the DisplayAllMessagesActivity

jonholmberg committed:
Cleaned up and changed DisplayAllMessagesActivity layout a bit

Cleaned up

Changed DisplayAllMessagesActivity layout a bit

Cleaned up

jonholmberg commented on jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/MainActivity.java:
```diff
@@ -127,6 +175,8 @@ public void showHide(Boolean showDisplayMessageFragment) {
         FragmentTransaction ft = fragmentManager.beginTransaction();
         ft.setCustomAnimations(android.R.anim.fade_in, android.R.anim.fade_out);
 
+        int displayOrientation = getResources().getConfiguration().orientation;
```
Don't know why I moved it up here - will move it back.

AndersHqst commented:
I don't understand "The file is deleted when the app is uninstalled, so that you can see messages from previous sessions." :)

AndersHqst commented on jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/DisplayAllMessagesActivity.java:
```diff
@@ -0,0 +1,57 @@
+
+public class DisplayAllMessagesActivity extends AppCompatActivity {
+
+    private String messageFileName;
```
This should be declared `static`

jonholmberg commented:
@AndersHqst Fair, that was bad phrasing. The point was that the file will remain on internal storage when the app is closed, so that you can see messages you wrote last time you had the app open, and it wont be deleted until the app is uninstalled.

AndersHqst commented on jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/MainActivity.java:
```diff
@@ -96,11 +118,37 @@ public void sendMessage(String message) {
+                    "Trying to send message, but displayMessageFragment is null.");
+        }
+
+        if (messageFile == null) {
```
Fine guard, but it should be placed in the `writeMessageToFile` method. Otherwise we would have to remember (which we won't) to do this check every time we call `writeMessageToFile`

jonholmberg replied:
True!

jonholmberg replied:
Actually, shouldn't it be 
```java
if (messageFile.exists()) {
  // write message
}
```
or (golden rule ?)
```java
if (!messageFile.exists()) {
  // log + return
}
// write message
```
or something similar?
We don't actually use the `messageFile` reference in the `writeMessageToFile` method, we just want to make sure the file is actually created before we write to it.

AndersHqst replied:
Use golden rule/path. That the file does not exist is the __exceptional case__ that we __guard__ against. Should the file for whatever exceptional reason not exist, we just return. For the rest of the method we can then just focus on the interesting part of the code, without being in the context of an `if`.

Eventhough `messageFile` isn't used in `writeMessageToFile`, the  "we just want to make sure the file is actually created before we write to it" is what we want a guard for, before doing the business in `writeMessageToFile`. If the check `messageFile == null` is a precondition for `writeMessageToFile`, we should put it in `writeMessageToFile`, or we will have to remember to do that check every time.

As an alternative, we could consider using the check to create the file if it is not there.

jonholmberg replied:
Yeah, I understand we have to move the check to `writeMessageToFile`and guard against the exceptional case that the file does not exist, **but** is the check `messageFile == null` not redundant even in `writeMessageToFile`? The code in `writeMessageToFile` doesn't depend on the `messageFile` reference (it just uses to filename to create an output stream).

As I'm looking into it, it actually looks like [openFileOutput](https://developer.android.com/reference/android/content/Context.html#openFileOutput(java.lang.String,int)) creates a file if it doesn't already exist, so that the `messageFile` reference is not needed at all? 

Tell me if I'm just confused and looking at it completely in the wrong way, but it seems like the `openFileOutput` handles all our problems for us?

AndersHqst replied:
I agree that the check `messageFile == null` probably does not belong in this part of your code.

Yes, I think you will be good with simply removing this. :)

AndersHqst commented on jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/DisplayMessageFragment.java:
```diff
@@ -15,8 +17,12 @@
+
+    // The messages to be displayed.
     private String message = "";
+    private String prev_message = "";
```
Use camelCasing

tudorgk replied:
🐫

AndersHqst commented on jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/DisplayMessageFragment.java:
```diff
@@ -25,6 +31,12 @@
         super.onCreate(savedInstanceState);
         setRetainInstance(true);
+
+        String no_prev_message = getResources().getString(R.string.no_prev_message);
```
Use camelCasing

tudorgk replied:
🐫

jonholmberg replied:
Wondering: didn't you say that values/variables created "on the go" should be `snake_case`? Or am I remembering it incorrectly / misunderstanding? :)

AndersHqst replied:
Not sure, I have probably been unclear. The answer is: "it doesn't matter, as long as you are consistent". So, we use `camelCase` for variable names, which is the most common in Java/Kotlin/Android. 

There is an exception to the above, where references to UI instances are set for us automatically. This adds instance variables generated from e.g. `R.id.text_view` for us, and those will exceptionally be in 🐍  case as e.g. `text_view`

jonholmberg replied:
Alright - cool! :)

AndersHqst commented on jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/DisplayMessageFragment.java:
```diff
@@ -40,6 +52,20 @@ public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
         display_message = (TextView) getView().findViewById(R.id.text_view_display_message);
         display_message.setText(message);
+        // Set the previous message.
+        TextView display_prev_message = (TextView) getView().findViewById(R.id.text_view_display_prev_message);
```
I would also use camelCasing here. As we have talked a bit about, we will however see this mixed in when e.g. `TextView` references are automatically generated for us. You will see this later.

tudorgk replied:
🐫

AndersHqst commented on jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/DisplayMessageFragment.java:
```diff
@@ -40,6 +52,20 @@ public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
+    public void onStop() {
+        super.onStop();
+
+        // Save the current message.
```
Consider moving to a `saveCurrentMessage` method

AndersHqst commented on jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/MainFragment.java:
```diff
@@ -58,22 +61,28 @@ public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
             delegate.sendMessage(editMessage.getText().toString());
-        } else {
+            editMessage.setText("");
+        } else if (v == showHideButton) {
```
Try always to be explicit and say `view` instead of `v`, to make the code self-explaining without us having to translate names in our heads.

jonholmberg replied:
Yes, sir!

AndersHqst commented on jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/MainFragment.java:
```diff
@@ -58,22 +61,28 @@ public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
         // Set listeners
         showHideButton.setOnClickListener(this);
         sendButton.setOnClickListener(this);
+        showAllMessagesButton.setOnClickListener(this);
```
An alternative to this is using an anonymous function: 
```
showAllMessagesButton.setOnClickListener(new View.OnClickListener() {
    @Override
    public void onClick(View view) {
        // Button was clicked
    }
});
```

AndersHqst replied:
This approach is often preferred as it avoids the stuff you have now in `onClick` which becomes more and more cluttered as we add more listeners

jonholmberg replied:
Nice! 🎉

jonholmberg committed:
Now writing to a persistent file in the internal storage

Minor fixes

Cleaning up. Now writing to a persistent file instead of to a temporary file

AndersHqst reviewed (commented):
Good stuff. Great to see you make proper use of access modifiers. I was a bit thrown off by the passing around of the filename in the intent extras. Is the idea that `DisplayAllMessagesActivity` should be generic and able to display text from any filename passed to it? Otherwise, the file name could just be one `public final static` that everyone references. It is fine as is.

Alternatively, as the file essentially is a database, you would put all your file read/write/nameing into some `FileAccessor` class. This greatly clean up your activities from all the file-handling code.

tudorgk commented on jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/DisplayMessageFragment.java:
```diff
@@ -25,6 +31,12 @@
+
+        // Get the previous message.
+        SharedPreferences sharedPreferences = getActivity().getPreferences(Context.MODE_PRIVATE);
+        prev_message = sharedPreferences.getString(PREV_MESSAGE, no_prev_message);
```
🐫

tudorgk commented on jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/DisplayMessageFragment.java:
```diff
@@ -15,8 +17,12 @@
+    private String prev_message = "";
 
     // The TextView that displays the message.
     private TextView display_message;
```
🐫

tudorgk commented on jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/MainActivity.java:
```diff
@@ -153,4 +201,13 @@ public void showHide(Boolean showDisplayMessageFragment) {
+    /** Creates a new activity that shows all the previous messages, i.e. the contents of the
+     * file containing the previously written messages. */
+    public void showAllMessages() {
+
```
Remove empty line.

tudorgk commented on jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/MainActivity.java:
```diff
@@ -35,11 +45,23 @@
         setupFragments();
+
+        try {
+            messageFile = File.createTempFile(messageFilename, null, this.getCacheDir());
```
Why are you saving persistent data as a cache file in the cache directory? What happens if the user deletes the cache folder?

jonholmberg replied:
Fair point - didn't think that through. All the messages would get deleted. Will change it so the messages are saved in a persistent file on the internal storage.

tudorgk commented on jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/DisplayAllMessagesActivity.java:
```diff
@@ -0,0 +1,57 @@
+    private String getAllMessages() {
+        FileInputStream messageInputStream;
+        int tempByte;
+        StringBuffer allMessages = new StringBuffer("");
```
Consider replacing `StringBuffer` with `StringBuilder`.

jonholmberg replied:
Ah yeah!

AndersHqst replied:
In case you wonder what the difference is (like me): http://stackoverflow.com/questions/355089/difference-between-stringbuilder-and-stringbuffer

AndersHqst replied:
Which says that we should always just use `StringBuilder`

jonholmberg replied:
Yeah I found that link as well 👍

tudorgk commented on jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/DisplayAllMessagesActivity.java:
```diff
@@ -0,0 +1,57 @@
+
+        try {
+            messageInputStream = openFileInput(messageFileName);
+            while ((tempByte = messageInputStream.read()) != -1) {
```
Is this really the way the tutorial suggest on doing it? Seems super annoying 😄

jonholmberg replied:
Yup it is very annoying :D They don't show a way to do it in the tutorial. There is a way to read multiple bytes at once using `FileInputStream.read(byte[] b)` (see [here](https://www.tutorialspoint.com/java/io/fileinputstream_read_byte.htm)), but I found that it sometimes (I think at the end of the file) it read some bytes twice! I couldn't figure out why, so I found two possibilities where the above is one of them. The other one is to check how many bytes are available and read accordingly. Both seem not so nice.. if you prefer the second option or if you have a way to do it nicer, please let me know :)

tudorgk reviewed (commented):
Adding my 2 cents here.

jonholmberg commented:
@AndersHqst You're right that currently there is no reason not to make the filename a `public final static` since `DisplayAllMessagesActivity` is only used to display the contents of the `messageFile`. I think I will change it to this. Make sense as well when considering the name of the class ;)

I see. So you want me to create a `FileAccessor` class with methods that handle all this nasty IO?

jonholmberg committed:
Added anonymous OnClickListeners and made the messageFilename public

Added anonymous OnClickListeners to clean up the onClick handling

Removed redundant null check and set the messageFileName to public

AndersHqst commented:
@jonholmberg Regarding the filename as a `public final static` and the idea of a `FileAccessor` class, don't worry about that, and just move on to more new stuff. Just wanted to consider the alternatives.

AndersHqst commented:
@jonholmberg 👍 from my side. Go ahead and merge when you feel you are done :)
